### ISSUE TITLE
Status of __builtins__ is not totally clear

### ISSUE THREAD

**Issue by b5f929de-c21c-4343-aebd-756fd1343a4c** [python/cpython#74557]
BPO | [30372](https://bugs.python.org/issue30372)
--- | :---
Nosy | @brettcannon, @terryjreedy, @mhsmith, @Mariatta, @nceder
PRs | <li>python/cpython#1725</li><li>python/cpython#1727</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-05-22.21:37:09.966>
created_at = <Date 2017-05-15.16:20:38.361>
labels = ['type-bug', '3.7', 'docs']
title = 'Status of __builtins__ is not totally clear'
updated_at = <Date 2017-05-22.21:37:09.965>
user = 'https://github.com/mhsmith'
```

bugs.python.org fields:
```python
activity = <Date 2017-05-22.21:37:09.965>
actor = 'Mariatta'
assignee = 'docs@python'
closed = True
closed_date = <Date 2017-05-22.21:37:09.966>
closer = 'Mariatta'
components = ['Documentation']
creation = <Date 2017-05-15.16:20:38.361>
creator = 'Malcolm Smith'
dependencies = []
files = []
hgrepos = []
issue_num = 30372
keywords = []
message_count = 7.0
messages = ['293722', '293773', '293972', '294181', '294183', '294188', '294189']
nosy_count = 6.0
nosy_names = ['brett.cannon', 'terry.reedy', 'docs@python', 'Malcolm Smith', 'Mariatta', 'NaomiCeder']
pr_nums = ['1725', '1727']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue30372'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by b5f929de-c21c-4343-aebd-756fd1343a4c** [comment]
https://docs.python.org/3.6/reference/executionmodel.html#builtins-and-restricted-execution describes the various things you can do with \_\_builtins__, but then says "Users should not touch \_\_builtins__; it is strictly an implementation detail." If this is so, the entire section should be marked "CPython implementation detail", not just that last paragraph.

Elsewhere, https://docs.python.org/3.6/reference/import.html#replacing-the-standard-import-system suggests that the \_\_import__ function could be replaced "at the module level to only alter the behaviour of import statements within that module". Is there any way of doing this other than manipulating the module's \_\_builtins__? If not, this is encouraging the programmer to rely on a CPython implementation detail without marking it as such.

**Comment by brettcannon** [comment]
You're correct that both are CPython implementation details and could be more clearly marked as such.

(In case anyone is curious, the code to look up \_\_import__() can be found at https://github.com/python/cpython/blob/c7ac7280c321b3c1679fe5f657a6be0f86adf173/Python/ceval.c#L4991 .)

**Comment by terryjreedy** [comment]
Moving "CPython implementation detail: " up a paragraph should be trivial.  Resolving the conflict between " \_\_builtins__ can be set to a user-created dictionary to create a weak form of restricted execution." and "Users should not touch \_\_builtins__;" seems less so.  I suspect the the latter was added well after the former, and that the former should have been deleted at that time.

**Comment by nceder** [comment]
Created a PR on this... it seems that making the Implementation Detail warning come first and deleting the limited execution sentence achieves what's needed here.

**Comment by Mariatta** [comment]
New changeset 43c8a9e39bb2d4f25ed7ae17a55f34e375d1868b by Mariatta (Naomi Ceder) in branch 'master':
bpo-30372: Clarify that '\_\_builtins__' is a CPython Implementation detail (GH-1725)
https://github.com/python/cpython/commit/43c8a9e39bb2d4f25ed7ae17a55f34e375d1868b

**Comment by Mariatta** [comment]
New changeset 3d3b189f8ef321ccfb144a46d50892939d6004aa by Mariatta in branch '3.6':
[3.6] bpo-30372: Clarify that '\_\_builtins__' is a CPython Implementation detail (GH-1725) (GH-1727)
https://github.com/python/cpython/commit/3d3b189f8ef321ccfb144a46d50892939d6004aa

**Comment by Mariatta** [comment]
PR merged, and I backport it to 3.6

Thanks everyone :)
